lib.rs

```diff
@@ -0,0 +1,511 @@
+
+            // check allowance
+            if self.allowance(token_in, caller, this) < amount_token_in {
+                return Err(AmmError::InsufficientAllowanceOf(token_in));
+            }
+
+            let amount_token_out = self.out_given_in(token_in, token_out, amount_token_in)?;
+
+            if amount_token_out < min_amount_token_out {
+                return Err(AmmError::TooMuchSlippage);
+            }
+
+            self.transfer_from_tx(token_in, caller, this, amount_token_in)?;
+            self.transfer_tx(token_out, caller, amount_token_out)?;
+
+            self.env().emit_event(Swapped {
+                caller,
+                token_in,
+                token_out,
+                amount_in: amount_token_in,
+                amount_out: amount_token_out,
+            });
+
+            Ok(())
+        }
+
+        /// Allows authorized accounts to withdraw liquidity from the AMM pool.
+        ///
+        /// This function enables designated accounts (typically admins) to remove liquidity from the pool.
+        /// It takes a vector of tuples specifying the token contract `account_id` (`token_out`) and the amount (`amount`)
+        /// to be withdrawn for each token.
+        ///
+        /// Requirements:
+        ///  * The caller must be authorized to perform withdrawals (usually the admin or authorized liquidity providers).
+        ///
+        /// # Arguments
+        /// * `withdrawals`: A vector of tuples specifying the token contract `account_id` and amount to withdraw for each token.
+        ///
+        /// Emits a `Withdrawn` event for each successful withdrawal.
+        ///
+        /// Throws PSP22Error on failed transactions
+        #[ink(message)]
+        pub fn withdrawal(
+            &mut self,
+            withdrawals: Vec<(AccountId, Balance)>,
+        ) -> Result<(), AmmError> {
+            let caller = self.env().caller();
+
+            // check role, under normal circumstances only designated account can remove liquidity
+            self.check_owner(caller)?;
+            withdrawals.into_iter().try_for_each(
+                |(token_out, amount)| -> Result<(), AmmError> {
+                    // transfer token_out from the contract to the caller
+                    self.transfer_tx(token_out, caller, amount)?;
+                    self.env().emit_event(Withdrawn {
+                        caller,
+                        token: token_out,
+                        amount,
+                    });
+
+                    Ok(())
+                },
+            )?;
+
+            Ok(())
+        }
+
+        /// Alters the swap_fee parameter
+        ///
+        /// Can only be called by the contract's Admin.
+        #[ink(message)]
+        pub fn set_swap_fee_percentage(
+            &mut self,
+            swap_fee_percentage: u128,
+        ) -> Result<(), AmmError> {
+            if swap_fee_percentage.gt(&100) {
+                return Err(AmmError::WrongParameterValue);
```

Comment:
Why not saturate it to 100 instead of throwing an error?

Reply:
In case it is an honest input mistake, the error would force one to enter a proper value. Setting the fee percentage at a potentially wrong value could cause financial losses and complaints from users